src/lsp/cobol_parser/expect/gen_expect.ml

```diff
@@ -0,0 +1,421 @@
+  |> pp_match_cases ppf
+    Fmt.(using Lr1.to_int int)
+    (pp_brackets @@ pp_list @@ pp_xsymbol_of_nt)
+    "[])"
+
+let emit_reducable_productions_in_env ppf =
+  Fmt.pf ppf {|
+let reducable_productions_in ~env : Menhir.production list =
+  List.rev_map Menhir.find_production @@@@
+  match Menhir.current_state_number env with
+|};
+  Lr1.fold (fun lr1 acc ->
+    match reducable_productions lr1 with
+    | [] -> acc
+    | defaults -> ProdMap.add_to_list defaults lr1 acc)
+  ProdMap.empty
+  |> inv (ProdMap.fold (fun s l acc -> (l, s)::acc)) []
+  |> pp_match_cases ppf
+    Fmt.(using Lr1.to_int int)
+    Fmt.(pp_brackets @@ pp_list @@ (using Production.to_int int))
+    "[]"
+
+let completion_entries_of =
+  let rec continue_while_terminal rhs i l =
+    match rhs.(i) with
+        | T t, _, _ when is_valid_for_comp t ->
+            continue_while_terminal rhs (i+1) (t::l)
+        | _ | exception Invalid_argument _ -> List.rev l
+  in
+  let next_tokens_from_item item =
+    let (prod, idx) = item in
+    let rhs = Production.rhs prod in
+    match rhs.(idx) with
+      | N nt, _, _ -> List.filter_map terminal_filter_map (Nonterminal.first nt)
+      | T _, _, _ -> begin
+        match continue_while_terminal rhs idx [] with
+          | [] -> []
+          | l -> [K (NEL.of_list l)] end
+      | exception Invalid_argument _ -> []
+  in
+  Lr1.tabulate begin fun lr1 ->
+    let custom_comp_entries = CompEntrySet.from_list_map
+      ~add:CompEntrySet.add_option
+      begin function
+        | N nonterm, _ -> nonterminal_filter_map nonterm
+        | _ -> None end
+      (Lr1.transitions lr1) in
+    CompEntrySet.from_list_map
+      ~init:custom_comp_entries ~add:CompEntrySet.add_list
+      next_tokens_from_item (Lr0.items @@ Lr1.lr0 lr1)
+  end
+
+let emit_acceptable_terminals_in_env ppf = (* taking transitions *)
+  Fmt.pf ppf {|
+let acceptable_terminals_in ~env : completion_entry list =
+  NEL.(match Menhir.current_state_number env with
+|};
+  Lr1.fold (fun lr1 acc ->
+    match completion_entries_of lr1 with
+      | s when CompEntrySet.is_empty s -> acc
+      | s -> CompEntryMap.add_to_list s lr1 acc)
+  CompEntryMap.empty
+  |> inv (CompEntryMap.fold (fun s l acc -> (l, CompEntrySet.elements s)::acc)) []
+  |> pp_match_cases ppf
+    Fmt.(using Lr1.to_int int)
+    (pp_brackets @@ pp_list pp_completion_entry)
+    "[])"
+
+let guess_default_value ~mangled_name ~typ =
+  match typ, mangled_name with
+  | _ when EzString.ends_with ~suffix:"option" typ ||
+          EzString.starts_with ~prefix:"option_" mangled_name
+  -> Some "None"
+  | _ when EzString.ends_with ~suffix:"list" typ ||
+  EzString.starts_with ~prefix:"loption_" mangled_name ||
+  EzString.starts_with ~prefix:"list_" mangled_name
+  -> Some "[]"
+  | _ when EzString.ends_with ~suffix:"unit" typ
+  -> Some "()"
+  | _ when EzString.ends_with ~suffix:"bool" typ ||
```

Comment:
It's probably an equality test that we want for `bool` and `unit`.

Why the `match` if you never desctruct a value (*i.e,* every pattern is `_`)?

Reply:
Just seemed cleaner than a series of `if ... then ... else`

Reply:
> Just seemed cleaner than a series of `if ... then ... else`

Sometimes it is. But here it's misleading. If you align them properly (with line breaks (only) right before every `then` and `else`), that is usually readable and avoids the clutter of wildcard patterns.